File: code/structural/model_solver.jl
# Model Solver for the Two-Type Vickrey Auction Model
# =====================================================

using Printf
# Closed-form equilibrium price and model-predicted moments for the
# consumption-collector wine auction model.
#
# Reference: quality_reports/theory_model_2026-03-29.md
#
# Parameters (theta): (alpha_c, v_bar_k, alpha_k, gamma, delta, lam_bar)
#   alpha_c  — consumption-type valuation scale
#   v_bar_k  — collector-type baseline valuation
#   alpha_k  — collector-type age sensitivity
#   gamma    — shape of consumption quality function h(a*)
#   delta    — rate of consumption-type exit post-maturity
#   lam_bar  — fraction of consumption-type bidders at maturity

# ---------------------------------------------------------------------------
# Segment definitions (normalised age a* = auction_age / maturity_date)
# ---------------------------------------------------------------------------

const SEGMENT_MIDPOINTS = Dict{String,Float64}(
    "young"   => 0.30,
    "approach" => 0.80,
    "peak"    => 1.30,
    "trough"  => 1.80,
    "antique" => 2.50,
)

const SEGMENT_ORDER = ["approach", "peak", "trough", "antique"]
const YOUNG_BIN = (0.0, 0.4)  # baseline bin boundaries

# ---------------------------------------------------------------------------
# Primitives
# ---------------------------------------------------------------------------

"""
    h_fn(a_star, gamma) -> Float64

Consumption quality function: h(a*) = (a*)^γ · exp(γ·(1 − a*))
Properties: h(0)=0, h(1)=1, increasing on [0,1], decreasing on (1,∞).
"""
function h_fn(a_star::Float64, gamma::Float64)::Float64
    a_star <= 0.0 && return 0.0
    (a_star^gamma) * exp(gamma * (1.0 - a_star))
end

"""
    vc_fn(a_star, alpha_c, gamma) -> Float64
Consumption-type valuation: v_c(a*) = α_c · h(a*)
"""
function vc_fn(a_star::Float64, alpha_c::Float64, gamma::Float64)::Float64
    alpha_c * h_fn(a_star, gamma)
end

"""
    vk_fn(a_star, v_bar_k, alpha_k) -> Float64
Collector-type valuation: v_k(a*) = v̄_k + α_k · a*
"""
function vk_fn(a_star::Float64, v_bar_k::Float64, alpha_k::Float64)::Float64
    v_bar_k + alpha_k * a_star
end

"""
    lam_fn(a_star, lam_bar, delta) -> Float64
Fraction of consumption-type bidders:
  λ(a*) = λ̄                           if a* ≤ 1
  λ(a*) = λ̄ · exp(−δ·(a* − 1))       if a* > 1
"""
function lam_fn(a_star::Float64, lam_bar::Float64, delta::Float64)::Float64
    a_star <= 1.0 && return lam_bar
    lam_bar * exp(-delta * (a_star - 1.0))
end

# ---------------------------------------------------------------------------
# Equilibrium price (expected second-order statistic)
# ---------------------------------------------------------------------------

"""
    price(a_star, theta, n) -> Float64

Expected second-highest bid from n i.i.d. draws of the two-point valuation
distribution {v_high with prob p_high, v_low with prob 1−p_high}.

E[Y_{(n-1:n)}] = v_high · Pr(≥2 are v_high) + v_low · Pr(≤1 is v_high)
             = v_high · (1 − prob_leq1) + v_low · prob_leq1
where
  prob_leq1 = (1−p_high)^n + n·p_high·(1−p_high)^(n−1)
"""
function price(a_star::Float64, theta::Vector{Float64}, n::Int=5)::Float64
    alpha_c, v_bar_k, alpha_k, gamma, delta, lam_bar = theta

    vc = vc_fn(a_star, alpha_c, gamma)
    vk = vk_fn(a_star, v_bar_k, alpha_k)
    la = lam_fn(a_star, lam_bar, delta)

    if vc >= vk
        v_high, v_low, p_high = vc, vk, la
    else
        v_high, v_low, p_high = vk, vc, 1.0 - la
    end

    prob_leq1 = (1.0 - p_high)^n + n * p_high * (1.0 - p_high)^(n - 1)
    v_high * (1.0 - prob_leq1) + v_low * prob_leq1
end

# ---------------------------------------------------------------------------
# Bin-averaged log price (replaces midpoint approximation)
# ---------------------------------------------------------------------------

"""
    avg_log_price_bin(a_lo, a_hi, theta, n, n_points) -> Float64

Average of log P(a*) over a uniform grid on [a_lo, a_hi].
Returns 1e6 (penalty) if any price is non-positive.
Eliminates Jensen's-inequality bias from single-midpoint approximation.
"""
function avg_log_price_bin(a_lo::Float64, a_hi::Float64,
                           theta::Vector{Float64}, n::Int, n_points::Int=20)::Float64
    step = (a_hi - a_lo) / (n_points - 1)
    s = 0.0
    @inbounds for i in 0:(n_points - 1)
        a = a_lo + i * step
        p = price(a, theta, n)
        p <= 0.0 && return 1e6
        s += log(p)
    end
    s / n_points
end

# ---------------------------------------------------------------------------
# Model-predicted moments
# ---------------------------------------------------------------------------

"""
    moments_integrated(theta, bins, n, n_points) -> Vector{Float64}

Compute model-predicted log-price differences relative to the young baseline bin.

bins: Vector of (label, lo, hi) tuples for the non-baseline bins.

Returns log P̄(bin_i) − log P̄(young) for each bin.
Returns fill(1e6, ...) if the baseline price is non-positive (penalty region).
"""
function moments_integrated(theta::Vector{Float64},
                            bins::Vector{Tuple{String,Float64,Float64}},
                            n::Int=5,
                            n_points::Int=20)::Vector{Float64}
    log_p_young = avg_log_price_bin(YOUNG_BIN[1], YOUNG_BIN[2], theta, n, n_points)
    log_p_young >= 1e5 && return fill(1e6, length(bins))

    m = Vector{Float64}(undef, length(bins))
    @inbounds for (i, (_, a_lo, a_hi)) in enumerate(bins)
        avg = avg_log_price_bin(a_lo, a_hi, theta, n, n_points)
        m[i] = avg - log_p_young
    end
    m
end

# ---------------------------------------------------------------------------
# Diagnostics
# ---------------------------------------------------------------------------

"""
    print_valuations(theta, n)
Print valuation decomposition at each segment midpoint.
"""
function print_valuations(theta::Vector{Float64}, n::Int=5)
    alpha_c, v_bar_k, alpha_k, gamma, delta, lam_bar = theta
    p0 = price(SEGMENT_MIDPOINTS["young"], theta, n)
    log_p0 = p0 > 0 ? log(p0) : NaN

    println(@sprintf("%-10s %5s %8s %8s %8s %8s %12s",
                     "Segment", "a*", "v_c", "v_k", "lambda", "Price", "log P-log P0"))
    println("-" ^ 65)
    for seg in vcat(["young"], SEGMENT_ORDER)
        a    = SEGMENT_MIDPOINTS[seg]
        vc   = vc_fn(a, alpha_c, gamma)
        vk   = vk_fn(a, v_bar_k, alpha_k)
        la   = lam_fn(a, lam_bar, delta)
        pval = price(a, theta, n)
        ld   = (pval > 0 && p0 > 0) ? log(pval) - log_p0 : NaN
        println(@sprintf("%-10s %5.2f %8.4f %8.4f %8.4f %8.4f %12.4f",
                         seg, a, vc, vk, la, pval, ld))
    end
end

File: code/structural/model_solver.py
"""
Model Solver for the Two-Type Vickrey Auction Model
====================================================

Closed-form equilibrium price and model-predicted moments for the
consumption-collector wine auction model.

Reference: quality_reports/theory_model_2026-03-29.md
"""

from pathlib import Path
import numpy as np
from scipy.integrate import quad

ROOT = Path(__file__).resolve().parent.parent.parent

# ---------------------------------------------------------------------------
# Segment definitions (normalised age a*)
# ---------------------------------------------------------------------------
SEGMENT_MIDPOINTS = {
    "young": 0.30,
    "approach": 0.80,
    "peak": 1.30,
    "trough": 1.80,
    "antique": 2.50,
}

SEGMENT_ORDER = ["approach", "peak", "trough", "antique"]


# ---------------------------------------------------------------------------
# Primitives
# ---------------------------------------------------------------------------

def h(a_star: float, gamma: float) -> float:
    """Consumption quality function.

    h(a*) = (a*)^gamma * exp(gamma * (1 - a*))

    Properties: h(0)=0, h(1)=1, increasing on [0,1], decreasing on (1,inf).
    """
    if a_star <= 0.0:
        return 0.0
    return (a_star ** gamma) * np.exp(gamma * (1.0 - a_star))


def v_c(a_star: float, alpha_c: float, gamma: float) -> float:
    """Consumption-type valuation: v_c(a*) = alpha_c * h(a*)."""
    return alpha_c * h(a_star, gamma)


def v_k(a_star: float, v_bar_k: float, alpha_k: float) -> float:
    """Collector-type valuation: v_k(a*) = v_bar_k + alpha_k * a*."""
    return v_bar_k + alpha_k * a_star


def lam(a_star: float, lam_bar: float, delta: float) -> float:
    """Probability that a bidder is consumption type.

    lambda(a*) = lam_bar                           if a* <= 1
    lambda(a*) = lam_bar * exp(-delta*(a* - 1))    if a* > 1
    """
    if a_star <= 1.0:
        return lam_bar
    return lam_bar * np.exp(-delta * (a_star - 1.0))


# ---------------------------------------------------------------------------
# Equilibrium price (expected second-order statistic)
# ---------------------------------------------------------------------------

def price(a_star: float, theta: np.ndarray, n: int = 5) -> float:
    """Equilibrium price: expected second-order statistic of n i.i.d. draws
    from the two-point valuation distribution.

    Parameters
    ----------
    a_star : float
        Normalised age (age / maturity date).
    theta : array-like, shape (6,)
        (alpha_c, v_bar_k, alpha_k, gamma, delta, lam_bar)
    n : int
        Number of bidders (>= 2).

    Returns
    -------
    float
        Expected second-highest bid (= equilibrium price).
    """
    alpha_c, v_bar_k, alpha_k, gamma, delta, lam_bar = theta

    vc = v_c(a_star, alpha_c, gamma)
    vk = v_k(a_star, v_bar_k, alpha_k)
    la = lam(a_star, lam_bar, delta)

    # Probability weights for the second-order statistic of a two-point
    # distribution with values {v_high, v_low} and probabilities {p_high, p_low}.
    #
    # E[Y_{(n-1:n)}] = v_high * Pr(at least 2 draws are v_high)
    #                 + v_low  * Pr(at most 1 draw is v_high)
    #
    # Pr(at most 1 draw = v_high) = (1-p_high)^n + n * p_high * (1-p_high)^(n-1)

    if vc >= vk:
        # v_high = vc with probability la
        p_high = la
        prob_at_most_1_high = (1.0 - p_high) ** n + n * p_high * (1.0 - p_high) ** (n - 1)
        p_val = vc * (1.0 - prob_at_most_1_high) + vk * prob_at_most_1_high
    else:
        # v_high = vk with probability (1-la)
        p_high = 1.0 - la
        prob_at_most_1_high = (1.0 - p_high) ** n + n * p_high * (1.0 - p_high) ** (n - 1)
        p_val = vk * (1.0 - prob_at_most_1_high) + vc * prob_at_most_1_high

    return p_val


def price_vec(a_star_arr: np.ndarray, theta: np.ndarray, n: int = 5) -> np.ndarray:
    """Vectorised price over an array of normalised ages."""
    return np.array([price(a, theta, n) for a in a_star_arr])


# ---------------------------------------------------------------------------
# Model-predicted moments
# ---------------------------------------------------------------------------

def moments(theta: np.ndarray, n: int = 5) -> np.ndarray:
    """Compute model-predicted log-price differences relative to young baseline.

    Returns
    -------
    np.ndarray, shape (4,)
        [log P(0.80) - log P(0.30),
         log P(1.30) - log P(0.30),
         log P(1.80) - log P(0.30),
         log P(2.50) - log P(0.30)]
    """
    p_young = price(SEGMENT_MIDPOINTS["young"], theta, n)

    if p_young <= 0.0:
        return np.full(4, 1e6)  # penalty for non-positive prices

    log_p_young = np.log(p_young)

    m = np.empty(4)
    for i, seg in enumerate(SEGMENT_ORDER):
        p_seg = price(SEGMENT_MIDPOINTS[seg], theta, n)
        if p_seg <= 0.0:
            m[i] = 1e6  # penalty
        else:
            m[i] = np.log(p_seg) - log_p_young

    return m


# ---------------------------------------------------------------------------
# Integrated moments (Fix: replaces midpoint approximation)
# ---------------------------------------------------------------------------

# Young baseline bin boundaries for normalisation
YOUNG_BIN = (0.0, 0.4)


def _avg_log_price_bin(a_lo: float, a_hi: float, theta: np.ndarray,
                       n: int, n_points: int = 20) -> float:
    """Average log P(a*) over [a_lo, a_hi] using Simpson's rule (n_points points).

    Returns large penalty value if any price is non-positive within the bin
    (prevents optimization from exploring degenerate regions).
    """
    a_grid = np.linspace(a_lo, a_hi, n_points)
    log_p_vals = np.empty(n_points)
    for j, a in enumerate(a_grid):
        p = price(a, theta, n)
        if p <= 0.0:
            return 1e6  # penalty — this parameter region is infeasible
        log_p_vals[j] = np.log(p)
    return float(np.mean(log_p_vals))  # trapezoidal rule on uniform grid


def moments_integrated(theta: np.ndarray,
                       bins: list,
                       n: int = 5,
                       n_points: int = 20) -> np.ndarray:
    """Compute model-predicted moments using bin-averaged log prices.

    Replaces single-midpoint approximation with average log P over each bin,
    eliminating Jensen's inequality bias and approximation error in wide bins.

    Parameters
    ----------
    theta : array-like, shape (6,)
        (alpha_c, v_bar_k, alpha_k, gamma, delta, lam_bar)
    bins : list of (label, a_lo, a_hi) or (a_lo, a_hi)
        Non-baseline bins. Each element is (label, lo, hi) or just (lo, hi).
    n : int
        Number of bidders.
    n_points : int
        Grid points per bin for numerical integration (default 20).

    Returns
    -------
    np.ndarray, shape (len(bins),)
        Average log P(a*) in each bin minus average log P in the young baseline.
    """
    # Young baseline: [0.0, 0.4)
    log_p_young = _avg_log_price_bin(YOUNG_BIN[0], YOUNG_BIN[1], theta, n, n_points)
    if log_p_young >= 1e5:
        return np.full(len(bins), 1e6)

    m = np.empty(len(bins))
    for i, b in enumerate(bins):
        # Accept either (label, lo, hi) or (lo, hi)
        if len(b) == 3:
            _, a_lo, a_hi = b
        else:
            a_lo, a_hi = b
        avg = _avg_log_price_bin(a_lo, a_hi, theta, n, n_points)
        m[i] = avg - log_p_young

    return m


# ---------------------------------------------------------------------------
# Diagnostics
# ---------------------------------------------------------------------------

def print_valuations(theta: np.ndarray, n: int = 5):
    """Print valuation decomposition at each segment midpoint."""
    alpha_c, v_bar_k, alpha_k, gamma, delta, lam_bar = theta

    print(f"{'Segment':<10} {'a*':>5} {'v_c':>8} {'v_k':>8} {'lambda':>8} {'Price':>8} {'log P-log P0':>12}")
    print("-" * 65)

    p0 = price(SEGMENT_MIDPOINTS["young"], theta, n)
    log_p0 = np.log(p0) if p0 > 0 else np.nan

    for seg_name in ["young"] + SEGMENT_ORDER:
        a = SEGMENT_MIDPOINTS[seg_name]
        vc_val = v_c(a, alpha_c, gamma)
        vk_val = v_k(a, v_bar_k, alpha_k)
        la_val = lam(a, lam_bar, delta)
        p_val = price(a, theta, n)
        log_diff = np.log(p_val) - log_p0 if p_val > 0 and p0 > 0 else np.nan
        print(f"{seg_name:<10} {a:5.2f} {vc_val:8.4f} {vk_val:8.4f} {la_val:8.4f} {p_val:8.4f} {log_diff:12.4f}")


if __name__ == "__main__":
    # Quick sanity check with plausible parameters
    # theta = (alpha_c, v_bar_k, alpha_k, gamma, delta, lam_bar)
    theta_test = np.array([2.0, 0.5, 0.3, 3.0, 1.5, 0.5])

    print("=== Model Solver Sanity Check ===\n")
    print(f"Parameters: alpha_c={theta_test[0]}, v_bar_k={theta_test[1]}, "
          f"alpha_k={theta_test[2]}, gamma={theta_test[3]}, "
          f"delta={theta_test[4]}, lam_bar={theta_test[5]}")
    print(f"n = 5\n")

    print_valuations(theta_test, n=5)

    print(f"\nModel moments: {moments(theta_test, n=5)}")
